build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const BAIL_OUT_CONDITIONS = {
+  // If this isn't a FunctionDeclaration, bail out on modification.
+  isNotFunction: (t, path) => !t.isFunctionDeclaration(path.node),
+
+  // If this FunctionDeclaration is a Generator, bail out on modification.
+  isGenerator: (t, path) => path.node.generator,
+
+  // If this FunctionDeclaration doesn't have a single return statement, bail out on modification
+  isNotSingleReturnStatment: (t, path) => {
+    if (!path.get('body').isBlockStatement()) {
+      return false;
+    }
+    const body = path.get('body.body.0');
+    return !body || !body.isReturnStatement();
+  },
+
+  // Since we don't know if exported members are 'new'd, bail out on modification.
+  isExported: (t, path) =>
+    path.findParent(
+      path =>
+        path.isExportNamedDeclaration() || path.isExportDefaultDeclaration()
+    ),
+
+  // If the function contains usage of `arguments`, bail out on modification.
+  containsArgumentsUsage: (t, path) => {
+    let containsArgumentsUsage = false;
+    path.traverse({
```

Comment:
Nit: Visitors like this should be hoisted out, since we have to expensive modifications before they're usable. Use a `state` param to encapsulate state:

```js
const visitor = {
  Identifier() {
    this.containsArgumentsUsage = true;
  }
};
function foo(path) {
  path.traverse(visitor, ) 
}

Reply:
Will take a look. Right now there isn't a significant performance concern with the transform and the clarity of how it operates is useful as we expand its capabilities.